File: 0394-decode-string/0394-decode-string.java
class Solution {
    
    private int i = 0;
    
    public String decodeString(String s) {
        
        StringBuilder sb = new StringBuilder();
        

        
        while(i < s.length() && s.charAt(i) != ']') {
            if (Character.isDigit(s.charAt(i))) {
                int k = 0;
                while ( i < s.length() && Character.isDigit(s.charAt(i))) {
                    k = k * 10 + (s.charAt(i++) - '0');
                }
                i++;
                String decoded = decodeString(s);
                i++;
                while (k-- > 0) {
                    sb.append(decoded);
                }
            }
            else {
                    sb.append(s.charAt(i++));
                }
        }
        return sb.toString();        
    }
}

File: 0394-decode-string/0394-decode-string.py
class Solution:
    def decodeString(self, s: str) -> str:
        
        stack = []
        currentNum = 0
        result = ''

        for char in s:
            if char.isdigit():
                currentNum = currentNum * 10 + int(char)
            else:
                if char == '[':
                    stack.append((result, currentNum))
                    result = ''
                    currentNum = 0
                elif char == ']':
                    prevStr, num = stack.pop()
                    result = prevStr + num * result
                else:
                    result += char

        return result